File: built-in plugins dsp/mb-analog-polysynth.cpp
/**
 * MB Analog Poly
 * Category : instrument
 * Type     : analog
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic analog polyphonic synthesizer with fat oscillators and resonant filters
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_ANALOG_POLYSYNTH_H
#define MB_ANALOG_POLYSYNTH_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbAnalogPolysynth : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-analog-polysynth";
    static constexpr const char* PLUGIN_NAME    = "MB Analog Poly";
    static constexpr const char* PLUGIN_TYPE    = "analog";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float osc1_shape = 0.5f;  // range [0, 1]
    float osc2_shape = 0.5f;  // range [0, 1]
    float osc_mix = 0.5f;  // range [0, 1]
    float filter_cutoff = 8000f;  // range [20, 20000]
    float filter_res = 0.3f;  // range [0, 1]
    float filter_env = 0.5f;  // range [0, 1]
    float lfo_rate = 2f;  // range [0.1, 20]
    float lfo_depth = 0.2f;  // range [0, 1]
    float unison = 1f;  // range [1, 8]
    float detune = 10f;  // range [0, 50]
    float volume = 0.8f;  // range [0, 1]
    };

    MbAnalogPolysynth() = default;
    ~MbAnalogPolysynth() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.osc1_shape = std::clamp(params.osc1_shape, 0f, 1f);
        params.osc2_shape = std::clamp(params.osc2_shape, 0f, 1f);
        params.osc_mix = std::clamp(params.osc_mix, 0f, 1f);
        params.filter_cutoff = std::clamp(params.filter_cutoff, 20f, 20000f);
        params.filter_res = std::clamp(params.filter_res, 0f, 1f);
        params.filter_env = std::clamp(params.filter_env, 0f, 1f);
        params.lfo_rate = std::clamp(params.lfo_rate, 0.1f, 20f);
        params.lfo_depth = std::clamp(params.lfo_depth, 0f, 1f);
        params.unison = std::clamp(params.unison, 1f, 8f);
        params.detune = std::clamp(params.detune, 0f, 50f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Analog Poly
        return input;
    }
};

#endif // MB_ANALOG_POLYSYNTH_H

File: built-in plugins dsp/mb-analog-polysynth.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbAnalogPolysynthPlugin: PluginDefinition = {
    id: 'mb-analog-polysynth',
    slug: 'mb-analog-polysynth',
    name: 'MB Analog Poly',
    category: 'instrument',
    type: 'analog',
    version: '1.0.0',
    description: 'Classic analog polyphonic synthesizer with fat oscillators and resonant filters',
    author: 'Max Booster', grade: 'A',
    oscillators: [
      { type: 'sawtooth', detune: -7, gain: 0.4 },
      { type: 'sawtooth', detune: 7, gain: 0.4 },
      { type: 'square', detune: 0, gain: 0.2 },
    ],
    envelope: { attack: 0.01, decay: 0.3, sustain: 0.7, release: 0.3 },
    parameters: [
      { id: 'osc1_shape', name: 'Osc 1 Shape', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'osc2_shape', name: 'Osc 2 Shape', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'osc_mix', name: 'Oscillator Mix', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'filter_cutoff', name: 'Filter Cutoff', type: 'float', defaultValue: 8000, minValue: 20, maxValue: 20000, unit: 'Hz', automatable: true },
      { id: 'filter_res', name: 'Filter Resonance', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true },
      { id: 'filter_env', name: 'Filter Envelope', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'lfo_rate', name: 'LFO Rate', type: 'float', defaultValue: 2, minValue: 0.1, maxValue: 20, unit: 'Hz', automatable: true },
      { id: 'lfo_depth', name: 'LFO Depth', type: 'float', defaultValue: 0.2, minValue: 0, maxValue: 1, automatable: true },
      { id: 'unison', name: 'Unison Voices', type: 'int', defaultValue: 1, minValue: 1, maxValue: 8, automatable: false },
      { id: 'detune', name: 'Unison Detune', type: 'float', defaultValue: 10, minValue: 0, maxValue: 50, unit: 'cents', automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { osc1_shape: 0.5, osc2_shape: 0.5, osc_mix: 0.5, filter_cutoff: 8000, filter_res: 0.3, filter_env: 0.5, lfo_rate: 2, lfo_depth: 0.2, unison: 1, detune: 10, volume: 0.8 },
  };

export default MbAnalogPolysynthPlugin;
